streaming/util/shorthand.py

```diff
@@ -1,115 +1,371 @@
 
-    Raises:
-        ValueError: Invalid byte suffix.
+    ret = {}
+    parts = text.split(sep)
+    for part in parts:
+        key, val = part.split(eq)
+        if key in ret:
+            raise ValueError(f'Repeated key: {key} (text: {text}).')
+        ret[key] = val
+    return ret
+
+
+def _normalize_arg(text: str, units: Dict[str, int], to_type: type) -> Union[int, float]:
+    """Normalize a human-friendly unit string to number.
+
+    Args:
+        text (str): Human-friendly string.
+        units (Dict[str, Any]): Mapping of unit name to value.
+        to_type (Union[int, float]): The return type.
 
     Returns:
-        int: Integer value of bytes.
+        type: Computer-friendly number.
     """
-    #input is already an int
-    if isinstance(bytes_str, int) or isinstance(bytes_str, float):
-        return int(bytes_str)
-
-    units = {
-        'kb': 1024,
-        'mb': 1024**2,
-        'gb': 1024**3,
-        'tb': 1024**4,
-        'pb': 1024**5,
-        'eb': 1024**6,
-        'zb': 1024**7,
-        'yb': 1024**8,
-    }
-    # Convert a various byte types to an integer
-    for suffix in units:
-        bytes_str = bytes_str.lower().strip()
-        if bytes_str.lower().endswith(suffix):
-            try:
-                return int(float(bytes_str[0:-len(suffix)]) * units[suffix])
-            except ValueError:
-                raise ValueError(''.join([
-                    f'Unsupported value/suffix {bytes_str}. Supported suffix are ',
-                    f'{["b"] + list(units.keys())}.'
-                ]))
+    # Must be non-empty.
+    if not text:
+        raise ValueError(f'Text is empty.')
+
+    # Drop commas and underscores (useful to demarcate thousands '1,337' or '1_337').
+    text = text.replace(',', '')
+    text = text.replace('_', '')
+
+    # Must start with a digit.
+    char = text[0]
+    if not char.isdigit():
+        raise ValueError(f'Text must start with a digit, but got {text[0]} instead (input: ' +
+                         f'{text}).')
+
+    # Must alternative between numbers and units, starting with a number.
+    in_num = True
+    part = []
+    parts = []
+    for char in text:
+        is_digit = char.isdigit() or char == '.'
+        if in_num:
+            if is_digit:
+                part.append(char)
+            else:
+                part = ''.join(part)
+                parts.append(part)
+                part = [char]
+                in_num = False
+        else:
+            if is_digit:
+                part = ''.join(part)
+                parts.append(part)
+                part = [char]
+                in_num = True
+            else:
+                part.append(char)
+    part = ''.join(part)
+    parts.append(part)
+
+    # If just a number, that's it.
+    if len(parts) == 1:
+        part, = parts
+        try:
+            return to_type(part)
+        except:
+            raise ValueError(f'Simple text must be numeric, but got {part} instead (input: ' +
```

Comment:
Same here. I find it difficult to understand what `Simple text` means here from a user perspective.

Also, looks at other exception messages in the PR for improvement.

Reply:
How about now...